# Capítulo 1

## Seção 1.1 -  The Elements of Programming

Based on the book: https://mitpress.mit.edu/sicp/full-text/book/book.html  

Additional material: https://www.cs.bgu.ac.il/~mira/ppl-book-full.pdf  

### Sintaxe:

- x                  :  Variável x
- ()                 :  Lista vazia; 
- (1 2 3)            :  Lista com 3 números; 
- ("Common" "Lisp")  :  Lista com duas strings; 
- (x y z)            :  Lista com três variáveis
- (x 1 "lisp")       :  Lista com uma variável, um número e uma string
- (+ (- 5 3) 4)      :  Lista com um variável, uma lista e um número

### Expressões

LISP é uma linguagem estruturada entre parênteses. Dessa forma as expressões são feitas da seguinte forma:
                                    
#### (< operador > < operandos >)

- Como operadores temos as expressões primitivas: +, *, - e /.
- Como operandos podemos ter dois ou mais elementos e esses elementos podem ser outras expressões.

Exemplos:

In [13]:
(+ 50 21)

71

In [9]:
(- 37 19)

18

In [10]:
(* 2 35)

70

In [11]:
(/ 36 3)

12

#### Método de Avaliação das Expressões:
- Busque a subexpressão mais interna à esquerda, aplique o operador aos operandos e faça o mesmo para a expressão à direita, se existir e depois aplique o operador da subexpressão que contém as subespressões avaliadas. Repita o método sempre da subexpressão mais interna à esquerda para fora.

Exemplo:


In [14]:
(+ (* 3 (+ (* 2 4) (+ 3 5))) (+ (- 10 7) 6))

57

No caso acima a avaliação se dará da seguinte forma:

1º Achamos a subexpressão mais interna à esquerda: (* 2 4) e aplicamos o operador aos operandos, resultando em 8;

2º Fazemos o mesmo para a subexpressão a sua direita, (+ 3 5), resultando em 8;

3º Aplicamos o operador (+) aos valores achados (8 e 8), resultando em 16;

4º Nesse momento, (* 3 (+ (* 2 4) (+ 3 5)) se reduziu a (* 3 16) que é 48;

5º Repita o que foi feito na subexpressão (+ (- 10 7) 6), o que resultará em 9;

6º Aplique o operador da expressão (+) aos dois operandos já reduzidos (48 e 9).

In [15]:
;Escrever dessa forma é a melhor maneira de entender a expressão e calcular seu valor.
(+ (* 3                 
      (+ (* 2 4)     
         (+ 3 5)))
   (+ (- 10 7)
      6))

57

Em Common Lisp nomeamos variáveis com setq e, temporariamente, com let.

In [16]:
(setq a 5)

SIMPLE-WARNING: 
undefined variable: A


5

In [18]:
(+ a 2)

7

In [22]:
(let ((a 2)) (+ a 1))

3

In [23]:
a

5

In [24]:
(* 2 a)

10

E definimos funções em Common Lisp com defun, segindo o modelo abaixo:
#### (defun < nome > (< parâmetros>) (< corpo >))

                       
Atenção: Os códigos do livro não estão em CL.

In [33]:
(defun hello-world ()
  (format t "hello, world"))

REDEFINITION-WITH-DEFUN: 
  redefining CL-JUPYTER-USER::HELLO-WORLD in DEFUN


HELLO-WORLD

In [34]:
(defun comprimento_circunferencia (r)
    (* 2 pi r))

COMPRIMENTO_CIRCUNFERENCIA

In [35]:
(defun area_circunf (r)
    (* pi (* r r)))

REDEFINITION-WITH-DEFUN: 
  redefining CL-JUPYTER-USER::AREA_CIRCUNF in DEFUN


AREA_CIRCUNF

In [36]:
(area_circunf 5)

78.53981633974483d0

In [37]:
(comprimento_circunferencia 5)

31.41592653589793d0

In [41]:
(defun square (x)
    (* x x))

(square 3)

9

In [38]:
(defun soma (x y)
    (+ x y))

(soma 3 4)

7

##### Modelo de Substituição:  
Outra forma de avaliar as expressões seria aplicar os procedimentos compostos, as funções, aos argumentos. Como fazer? Avalie o corpo da função trocando cada parâmetro formal pelo seu argumento correspondente. Veja o exemplo abaixo:

In [42]:
(defun sum-of-squares (x y)
    (+ (square x) (square y)))

REDEFINITION-WITH-DEFUN: 
  redefining CL-JUPYTER-USER::SUM-OF-SQUARES in DEFUN


SUM-OF-SQUARES

In [43]:
(sum-of-squares 3 4)

25

In [44]:
(defun f (a)
    (sum-of-squares (+ a 1) (* a 2)))
(f 5)

136

###### O que foi feito? 

Ao ser chamado, (f 5) chamou (sum-of-squares (+ 5 1) (* 5 2)) = (sum-of-squares 6 10) = (+ (square 6) (square 10)) = (+ (* 6 6) (* 10 10)) = (+ 36 100) = 136

#### Expressões Condicionais e Atributos:

* O construtor cond (que "significa" condição) é chamado de análise de caso, pois analizará os atributos na ordem dada, e, caso este seja verdadeiro, aplicará a expressão, seguindo o formato:
                                         
                                         (cond   ( <p1> <e1> )
                                                 ( <p2> <e2> )
                                                      ...
                                                 ( <pn> <en> )
                                                 

Veja o exemplo:


In [49]:
(defun abs-1 (x)
    (cond ((> x 0) x)
        ((eq x 0) 0)
        ((< x 0) (- x))))

REDEFINITION-WITH-DEFUN: 
  redefining CL-JUPYTER-USER::ABS-1 in DEFUN


ABS-1

In [50]:
(abs-1 8)

8

In [51]:
(abs-1 0)

0

In [52]:
(abs -1)

1

* A forma especial if é usada quando temos precisamente dois casos na análise de caso. Sua forma geral é:
                                  (if <atributo> <consequência> <alternativa>)
e é avaliada dada da seguinte forma: o < atributo > é avaliado e se for verdadeiro (True), faz-se o que está na < consequência > , caso seja falso (False), o faz-se o que está na < alternativa >. 

Exemplos:

In [56]:
(defun abs-2 (x)
    (if (>= x 0) x (- x)))

ABS-2

In [57]:
(abs-2 9)

9

In [58]:
(abs-2 0)

0

In [59]:
(abs-2 -5)

5

* Podemos, também, usar composições lógicas (e, ou, não) para formar os atributos:


####  - AND:                                    
                                      (and < e1> ... < en>)
O interpretador avalia as expressões < e > uma por uma e da esquerda para a direita. Se a avaliação de algum < e > for falso, a expressão toda é falsa e os outros < e > não são avaliadas. Se todos os < e > são verdadeiros, o valor da and expression é o valor do último. 

Exemplo:

In [60]:
(defun entre5e10 (x)
    (if (and (> x 5) (< x 10))
        t                          ; Retorna True: " o número está entre 5 e 10"
        nil))                      ; Quando queremos retornar false, retornamos nil: "o número não está entre 5 e 10"

ENTRE5E10

In [61]:
(entre5e10 7)

T

In [62]:
(entre5e10 3)

NIL

In [63]:
(entre5e10 10)

NIL

#### - OR:
                                            (or <e1> ... <en>) 
O interpretador avalia as expressões < e > uma por uma e da esquerda para a direita. Se a avaliação de algum < e > for verdadeiro, a expressão toda é verdadeira e os outros < e > não são avaliadas. Se todos os < e > são falsos, o valor da or expression é falsa.

Exemplo:

In [64]:
(defun maiorigual (x y)
    (if (or (> x y) (= x y))
        t
        nil))

MAIORIGUAL

In [65]:
(maiorigual 2 3)

NIL

In [66]:
(maiorigual 3 2)

T

#### - NOT:
                                        (not <e>)
O valor da not expression é verdadeiro quando a avaliação do < e > for falsa, e falso quando a avaliação do < e > for verdadeira.

Exemplo:

In [67]:
(defun maiorigual-2 (x y)
    (if (not (< x y))
        t
        nil))

MAIORIGUAL-2

In [68]:
(maiorigual-2 4 5)

NIL

In [69]:
(maiorigual-2 5 4)

T

#### - Exemplo: Raiz quadrada pelo método de Newton (Seção 1.1.7 - pág. 23)

O modo pelo qual o computador calcula raízes é usando o Método de Newton com sucessivas aproximações, que diz que sempre que tivermos uma suposição y para a raiz quadrada de x podemos fazer as seguintes manipulações para chegarmos a uma suposição tão perto quanto quisermos do valor real da raiz:
 - faça o quociente x/y
 - pegue esse valor e faça a média dele com o y: [(y+(x/y))/2]
 - esse valor será o seu novo y, repita a operação até que a diferença entre seu "novo y" e "velho y" seja tão pequeno quanto você queira.                                                       

In [70]:
(defun sqrt-iter (guess x)
    (if (good-enough? guess x)
        guess
        (sqrt-iter (improve guess x)
                   x)))

SIMPLE-STYLE-WARNING: 
  undefined function: GOOD-ENOUGH?

SIMPLE-STYLE-WARNING: 
  undefined function: IMPROVE


SQRT-ITER

In [71]:
(sqrt-iter 2.5 4)

UNDEFINED-FUNCTION: 
  #<UNDEFINED-FUNCTION GOOD-ENOUGH? {1004F3F073}>


NIL

In [72]:
(defun improve (guess x)
    (average guess (/ x guess)))

SIMPLE-STYLE-WARNING: 
  undefined function: AVERAGE


IMPROVE

In [73]:
(defun average (x y)
    (/ (+ x y) 2))

AVERAGE

In [74]:
(defun square (c)
    (* c c))

REDEFINITION-WITH-DEFUN: 
  redefining CL-JUPYTER-USER::SQUARE in DEFUN


SQUARE

In [75]:
(defun good-enough? (guess x)
    (< (abs (- (square guess) 
               x)
            0.001)))

SIMPLE-WARNING: 
The function was called with two arguments, but wants exactly one.


GOOD-ENOUGH?

In [76]:
(defun sqrt (x)
    (sqrt-iter 1.0 x))

REDEFINITION-WITH-DEFUN: 
  redefining COMMON-LISP:SQRT in DEFUN

SYMBOL-PACKAGE-LOCKED-ERROR: 
setting fdefinition of SQRT


NIL

In [78]:
(sqrt 123)

11.090536

In [79]:
(sqrt (+ 100 37))

11.7046995

#### Procedimentos como abstrações Caixa-preta (Seção 1.1.8 - pág. 26)

Tomando como exemplo a função good-enough?, quando a definimos em termos de square, não estávamos preocupados em como a função square computaria o quadrado do número dado, mas sim que ela computaria o valor. Dessa forma, square passou de uma função para a abstração de um procedimento, qualquer procedimento que computa o quadrado seria igualmente útil.

Levando em conta só o que a função retorna, temos duas formas de retornar o quadrado:

In [80]:
(defun square1 (c)
    (* c c))

SQUARE1

In [89]:
(defun square2 (x)
    (exp (double (log x))))

REDEFINITION-WITH-DEFUN: 
  redefining CL-JUPYTER-USER::SQUARE2 in DEFUN


SQUARE2

E duas formas de retornar o dobro de um número,

In [82]:
(defun double (x)
    (* 2 x))

DOUBLE

In [83]:
(defun double2 (x)
    (+ x x))

DOUBLE2

Dessa forma, a definição de um procedimento deve ser capaz de suprimir detalhes.

##### Nomes locais:
A escolha dos parâmetros deve ser independente do propósito do procedimento. Por exemplo, (defun double (x)(* 2 x)) deve ser a mesma coisa que (defun double (y) (* 2 y)), dessa forma eles serão parâmetros locais. Vamos ver que problemas podemos encontrar:

                            (defun good-enough? (guess x)
                                 (< (abs (- (square guess) 
                                             x)
                                          0.001)))
    
No exemplo acima temos a função square dentro da definição da função good-enough? e todas recebem a variável x como parâmetro de entrada. Se as variáveis não forem locais no corpo da função, o x na função square poderia ter sido confundido com o x da good-enough? e o comportamento da good-enough? dependeria da versão do square que usamos. Square não seria a caixa-preta.

Um parâmetro formal de um procedimento tem um papel muito especial na definição do procedimento, na medida em que não importa o nome do parâmetro formal. Esse nome é chamado de uma variável vinculada (bound variable), e dizemos que a definição de procedimento liga seus parâmetros formais. O significado de uma definição de procedimento é inalterado se uma variável vinculada for constantemente renomeada ao longo da definição. Se uma variável não está vinculada, dizemos que é free. O conjunto de expressões para o qual uma vinculação define um nome é chamado de escopo desse nome. Em uma definição de procedimento, as variáveis vinculadas declaradas como os parâmetros formais do procedimento têm o corpo do procedimento como seu escopo.

##### Estuturando as funções em blocos de funções:

Como vimos anteriormente, para definirmos a função sqrt tivemos que definir outras funções (sqrt-iter, good-enough? e improve). Entretanto, podemos fazer de outra forma, definindo blocos de funções:

In [90]:
(defun sqrt (x)
    (defun good-enough? (guess)
        (< (abs (- (square guess) x)) 0.001))
    (defun improve (guess)
        (average guess (/ x guess)))
    (defun sqrt-iter (guess)
        (if (good-enough? guess)
            guess
            (sqrt-iter (improve guess))))
    (sqrt-iter 1.0))


REDEFINITION-WITH-DEFUN: 
  redefining COMMON-LISP:SQRT in DEFUN

SYMBOL-PACKAGE-LOCKED-ERROR: 
setting fdefinition of SQRT


NIL

Vamos usar a block structure extensivamente para nos ajudar a dividir grandes programas em peças. A idéia de estrutura de blocos é originada pela linguagem de programação Algol 60. Aparece na maioria das linguagens de programação avançadas e é uma ferramenta importante para ajudar a organizar a construção de grandes programas.

#### Loops em CLisp

In [2]:
(setq a '(a b c d e))

SIMPLE-WARNING: 
undefined variable: A


(A B C D E)

In [3]:
(loop for x in a
      do (print x))


A 
B 
C 
D 
E 

SIMPLE-WARNING: 
undefined variable: A


NIL

In [6]:
(setq listavazia ())
listavazia

SIMPLE-WARNING: 
undefined variable: LISTAVAZIA


NIL

In [8]:
(setq novalista (append '(1 2 3) listavazia))
novalista

SIMPLE-WARNING: 
undefined variable: LISTAVAZIA

SIMPLE-WARNING: 
undefined variable: NOVALISTA


(1 2 3)

In [9]:
(setq novalista (append '(1 2 3) novalista))
novalista

SIMPLE-WARNING: 
undefined variable: NOVALISTA


(1 2 3 1 2 3)